raiden_contracts/data/source/services/ServiceRegistry.sol

```diff
@@ -3,49 +3,167 @@ pragma solidity 0.5.4;
 import "raiden/Token.sol";
 import "raiden/Utils.sol";
 
+contract Deposit {
+    // This contract holds ERC20 tokens as deposit until a predetemined point of time.
+
+    // The ERC20 token contract that the deposit is about.
+    Token public token;
+
+    // The address that can withdraw the deposit after the release time.
+    address public withdrawer;
+
+    // The timestamp after which the withdrawer can withdraw the deposit.
+    uint256 public release_at;
+
+    /// @param _token The address of the ERC20 token contract where the deposit is accounted.
+    /// @param _release_at The timestap after which the withdrawer can withdraw the deposit.
+    /// @param _withdrawer The address that can withdraw the deposit after the release time.
+    constructor(address _token, uint256 _release_at, address _withdrawer) public {
+        token = Token(_token);
+        // Don't care even if it's in the past.
+        release_at = _release_at;
+        withdrawer = _withdrawer;
+    }
+
+    // In order to make a deposit, transfer the ERC20 token into this contract.
+    // If you transfer a wrong kind of ERC20 token or ETH into this contract,
+    // these tokens will be lost forever.
+
+    /// @notice Withdraws the tokens that have been deposited.
+    /// Only `withdrawer` can call this.
+    /// @param _to The address where the withdrawn tokens should go.
+    function withdraw(address _to) external returns (bool success) {
+        uint256 balance = token.balanceOf(address(this));
+        require(msg.sender == withdrawer);
+        require(now >= release_at);
+        require(balance > 0);
+        require(token.transfer(_to, balance));
+        return true;
+    }
+}
+
+
 contract ServiceRegistry is Utils {
     Token public token;
-    address public owner;
 
-    mapping(address => uint256) public deposits;  // token amount staked by the service provider
+    // After a price is set to set_price at timestamp set_price_at,
+    // the price decays according to decayed_price().
+    uint256 public set_price;
+    uint256 public set_price_at;
+
+    // Once the price is at min_price, it can't decay further.
+    uint256 constant min_price = 1000;
+
+    mapping(address => uint256) public service_valid_till;
     mapping(address => string) public urls;  // URLs of services for HTTP access
-    address[] public service_addresses;  // list of available services (ethereum addresses)
+
+    event RegisteredService(address indexed service, uint256 valid_till, uint256 deposit_amount, Deposit deposit_contract);
 
     // @param _token_for_registration The address of the ERC20 token contract that services use for registration fees
-    constructor(address _token_for_registration) public {
+    constructor(address _token_for_registration, uint256 _initial_price) public {
```

Comment:
Should `min_price` be settable as well?

Reply:
This follows in the next PR.  `min_price`, `6/5` bumping ratio and `200 days` will be configurable.